arangod/Replication2/ReplicatedLog/Algorithms.cpp

```diff
@@ -326,29 +326,165 @@ auto algorithms::updateReplicatedLog(LogActionContext& ctx, ServerID const& serv
+  return std::find(std::begin(flags), std::end(flags), ParticipantFlag::Failed) != std::end(flags);
+};
 
-  if (spearhead == commitIndex) {
-    return {commitIndex, CommitFailReason::withNothingToCommit()};
+auto operator<=(ParticipantStateTuple left, ParticipantStateTuple right) noexcept -> bool {
+  if (left.index < right.index) {
+    return true;
+  } else if (left.index==right.index) {
+    return left.id <= right.id;
+  } else {
+    return false;
+  }
+}
+
+
+/* Add flag for failed server to IndexParticipant struct
+ *
+ *  add struct for parameters:
+ *   replicationFactor
+ *   - writeConcern (==quorumSize)
+ *   - softWriteConcern (better name?)
+ *   -
+ *  TRI_ASSERT(indexes.size() == replicationFactor)
+ *  TRI_ASSERT(writeConcern <= softWriteConcern <= replicationFactor)
+ *
+ *   actualWriteConcern = max(writeConcern, x)
+ *     where x = min(replicationFactor - number of failedServers,
+ *                   softWriteConcern);
+ *
+ *  for purposes of computing actualwriteConcern, excluded and failed
+ *  servers are treated the same.
+ *
+ * writeConcern is the *minimum* number of confirmed copies necessary to
+ * proceed with a commit for any given LogIndex
+ *
+ * softWriteConcern allows for a number of failed servers to be tolerated
+ * for the purposes of commit
+ *
+ * replicationFactor is the (expected) number of participants in the
+ * replicated log
+ *
+ * */
+
+algorithms::CalculateCommitIndexOptions::CalculateCommitIndexOptions(
+    std::size_t writeConcern, std::size_t softWriteConcern, std::size_t replicationFactor)
+  : _writeConcern(writeConcern),
+    _softWriteConcern(softWriteConcern),
+    _replicationFactor(replicationFactor) {
+  TRI_ASSERT(_writeConcern <= _softWriteConcern)
+      << "writeConcern > softWriteConcern " << _writeConcern << " > " << _softWriteConcern;
+  TRI_ASSERT(_softWriteConcern <= _replicationFactor)
+      << "softWriteConcern > opt.replicationFactor " << _softWriteConcern
+      << " > " << _replicationFactor;
+}
+
+auto algorithms::calculateCommitIndex(std::vector<ParticipantStateTuple> const& indexes,
+                                      CalculateCommitIndexOptions const opt,
+                                      LogIndex currentCommitIndex, LogIndex spearhead)
+    -> std::tuple<LogIndex, CommitFailReason, std::vector<ParticipantId>> {
+  TRI_ASSERT(indexes.size() == opt._replicationFactor)
+      << "number of participants != replicationFactor (" << indexes.size()
+      << " < " << opt._replicationFactor << ")";
+
+  // number of failed participants
+  auto nrFailed = std::count_if(std::begin(indexes), std::end(indexes),
+                                [](auto& p) { return p.isFailed(); });
+  auto actualWriteConcern =
+      std::max(opt._writeConcern,
+               std::min(opt._replicationFactor - nrFailed, opt._softWriteConcern));
+  // vector of participants that are neither excluded nor
+  // have failed
+  auto eligible = std::vector<ParticipantStateTuple>{};
```

Comment:
Because indexes is now const all it's elements have stable addresses, so eligible could just store references to the tuples.